### ISSUE TITLE
Support sparse arrays as PSparse

### ISSUE THREAD

**Issue by PallHaraldsson** [noahbenson/Air.jl#2]
Hi,

I don't need this personally, it's more of a question for now, not pushing for anything. This idea should work for sparse arrays, right? Got me thinking seeing this question:

https://github.com/JuliaLang/julia/issues/42921#issuecomment-959159667

Sparse arrays/vectors, are basically implemented by regular vectors (and you already support those) so it seems like a trivial change. And making dimensions larger, at least for 1D arrays (not for multi?). What about adding dimensions or dropping from an array, would your code handle or could easily?

**Comment by noahbenson** [comment]
Hello,

So `PArray`s are already sparse arrays or dense arrays, either one. The implementation of `PArray` is a persistent bitmapped trie, which are performant for both dense arrays and sparse arrays; the only difference is whether you require all elements in the array to be explicitly specified or whether you allow there to be a default value like `0`. In `Air`, if you create a `PArray` using the code `PArray{Int,1}(0, 1000000)` it doesn't actually allocate 1,000,000 elements; it just notes the array's size and default element, like in a sparse array. When you insert or update elements, it does so sparsely (but it is also efficient at storing many elements in a row because of how the trie data structure works). Basically, persistent arrays behave well as both and it's not really useful to make a distinction between dense and sparse arrays.

So TL;DR—to make `PSparse` you probably just want something like `const PSparse = PArray`.

I haven't added code yet that would, for example, extend an array with its default value, but it's not a bad idea. For now, pushing the default element (or using pushfirst) will extend the array one element at a time without allocating more space.

**Comment by PallHaraldsson** [comment]
I think the key from your docs is
>have a performance comparable to their clojure counterparts. 
[..]
            PArray, a persistent array type that mimics Julia's native Array
[..]
    Plans that are not yet implemented
        Improved Persistent Array Methods. Currently, most methods of persistent arrays yield mutable Julia arrays instead of persistent arrays.

I.e. they only "mimic", and I looked up for Clojure, there with Log32N indexing. I was hoping persistent [n-dim] arrays could be implemented somehow with direct indexing. Still great to have this package. I really liked the idea of persistent data structures when I first saw it (but also non-pure functional languages like Julia for mutable arrays, rather than lists are the basic data-structure), thinking arrays the fatal flaw. Or maybe not fatal, you do have them sort of, I was hoping something more array-like still possible. Have you looked at/heard of the freeze/thaw proposal to make arrays immutable in Julia? It seems that could help or be all you need. For Clojure I see rseq for vectors, i.e. (only?) reverse order, so do you also implement an iterator for PArray, that might be faster than indexing, or as fast, or close to, as iterating over a Julia Array?

I guess something like quicksort of PArray, is, well, not going to be "quick". Do you know if some other algorithm is faster for sorting (then which [fastest]?), or some workarounds (copying to a mutable array, then sorting it and copy back fastest?)? The good thing about it is that it's in-place (and fast).

I'm also thinking if the array is n-dim, or even just 2D, you might want to sort and then move a slice of a table around. I guess that would be more than one indexing for a row, i.e, times n columns, and bad for your data-structure. In such a case likely better to have a jagged array and only move a 64-bit pointer (as in Java, and I suppose also the only thing in Clojure?).

>extend an array with its default value, but it's not a bad idea. For now, pushing the default element (or using pushfirst) will extend the array one element at a time without allocating more space.

If you mean in the former case, indexing-out-of-bounds would extend (or read) an array, as in MATLAB, then please don't do that. For Julia arrays not done, on purpose, as would be a potential source for bugs (and slower). I like the latter, i.e. explicit, and as close you can do as in Julia already. Also it's not clear this is easy for 2+ dim array.

**Comment by noahbenson** [comment]
Hi Pall,
I'm having trouble making sense of some of your comments here.

> I.e. they only "mimic", and I looked up for Clojure, there with Log32N indexing. I was hoping persistent [n-dim] arrays could be implemented somehow with direct indexing.

I don't know what you mean here. The persistent arrays in `Air` are persistent arrays; they mimic julia arrays in that they have approximately the same API for functions like `push()` / `push!()` and `setindex()` / `setindex!()`. If you want arrays with direct indexing, like the immutable arrays proposal for Julia, you will end up with a data structure for which updates of individual elements, deletions, push, and pop operations, and many other standard functions take O(n) time. With the persistent arrays in Air (and those in Clojure), these operations are O(log n), which is why data structures like these are useful in functional programming paradigms—you can update a finite number of elements in a persistent array quickly without reallocating the whole thing.

> For Clojure I see rseq for vectors, i.e. (only?) reverse order, so do you also implement an iterator for PArray, that might be faster than indexing, or as fast, or close to, as iterating over a Julia Array?

Clojure has rseq and a lot of other ways to iterate over arrays, but I don't understand your point here. Iteration is supported for `PArray` and it's reasonably fast. In order for it to be as fast as iteration for standard Julia arrays, the `PArray` type would need to be much slower for operations like push and setindex.

> I guess something like quicksort of PArray, is, well, not going to be "quick". Do you know if some other algorithm is faster for sorting (then which [fastest]?), or some workarounds (copying to a mutable array, then sorting it and copy back fastest?)? The good thing about it is that it's in-place (and fast).

It sounds like you are talking about Julia's native arrays, not my persistent array class here. At least, this idea you propose of having rows moved by pointer is not possible in Julia's native arrays as far as I know. However, you are always welcome to store your arrays as either `Array{Array{T,N},1}` or `PArray{PArray{T,N},1}`, if you prefer to have non-contiguous arrays.

> I'm also thinking if the array is n-dim, or even just 2D, you might want to sort and then move a slice of a table around. I guess that would be more than one indexing for a row, i.e, times n columns, and bad for your data-structure. In such a case likely better to have a jagged array and only move a 64-bit pointer (as in Java, and I suppose also the only thing in Clojure?).

The idea of doing anything in-place on a persistent or immutable array is nonsensical. The arrays cannot change so nothing can be done in-place. Regardless this comment sounds like speculation about premature optimization. The `PArray` class is fast at converting an array to a `PArray`; if you are noticing that `PArray` is really slow when you sort it or when you convert a `PArray` to an `Array` then sort and convert it back, then please open an issue about it, but right now, I'm not sure what you're getting at. You want the persistent array type to support faster quicksort than it currently does? That's not really a priority for me, but if you want to write a fast quicksort algorithm for `PArray` and submit a pull request, I'll look it over!

> If you mean in the former case, indexing-out-of-bounds would extend (or read) an array, as in MATLAB, then please don't do that. For Julia arrays not done, on purpose, as would be a potential source for bugs (and slower). I like the latter, i.e. explicit, and as close you can do as in Julia already. Also it's not clear this is easy for 2+ dim array.

What I said (and what you quoted) was, "pushing the default element (or using pushfirst) will extend the array one element at a time without allocating more space". The `push()` (or `push!()`) and `pushfirst()` (or `pushfirst!()`) functions are supposed to append or prepend elements to an array, whether a native Julia array or a `PArray` object. If you make a `PArray` with a default element of 0, then you push a 0 onto the end of the array (which yields a copied array containing the additional 0), the new array will not actually allocate space for that 0 because it is the default value (and because `PArray`s are naturally also sparse arrays). I don't know where you got the bit about indexing out-of-bounds—this doesn't even make sense because, again, `PArray`s are persistent. Changing them when someone indexes out of bounds would break their persistence.

I'm having trouble following what you actually want or are suggesting here, and whether these are just idle thoughts about how you use arrays in your code or whether you have an actual desire to see certain behavior in `Air` based on some knowledge of how Julia and persistent data structures are generally used. As such, I'm going to close this issue, but if you can fill in the blanks for me, I'll reopen it.
